archive/walk.go

```diff
@@ -2,18 +2,232 @@ package archive
 
 import (
 	"context"
+	"fmt"
 	"io"
+	"path"
+	"regexp"
+	"sort"
+	"strconv"
 	"strings"
+	"time"
 
 	"github.com/brimsec/zq/driver"
+	"github.com/brimsec/zq/pkg/byteconv"
 	"github.com/brimsec/zq/pkg/iosrc"
+	"github.com/brimsec/zq/pkg/nano"
 	"github.com/brimsec/zq/scanner"
 	"github.com/brimsec/zq/zbuf"
 	"github.com/brimsec/zq/zio"
 	"github.com/brimsec/zq/zio/detector"
 	"github.com/brimsec/zq/zng/resolver"
+	"github.com/segmentio/ksuid"
 )
 
+const (
+	dataDirname = "zd"
+	zarExt      = ".zar"
+)
+
+// A fileKind is the first part of a file name, used to differentiate files
+// when they are listed from the archive's backing store.
+type fileKind string
+
+const (
+	fileKindUnknown fileKind = ""
+	fileKindData             = "d"
+	fileKindSeek             = "ts"
+)
+
+// A dataFile holds archive record data. Only one kind of data file
+// currently exists, representing a file created during ingest.
+type dataFile struct {
+	id   ksuid.KSUID
+	kind fileKind
+}
+
+func newDataFile() dataFile {
+	return dataFile{ksuid.New(), fileKindData}
+}
+
+func (f dataFile) name() string {
+	return fmt.Sprintf("%s-%s.zng", f.kind, f.id)
+}
+
+var dataFileNameRegex = regexp.MustCompile(`d-([0-9A-Za-z]{27}).zng$`)
+
+func dataFileNameMatch(s string) (f dataFile, ok bool) {
+	match := dataFileNameRegex.FindSubmatch([]byte(s))
+	if match == nil {
+		return
+	}
+	id, err := ksuid.Parse(byteconv.UnsafeString(match[1]))
```

Comment:
Nit: No need for []byte(s) or byteconv.UnsafeString if you use dataFileNameRegex.FindStringSubmatch.